File: Bit Manipulation/1863. Sum of All Subset XOR Totals.cpp
// 2025/04/05
class Solution {
    public:
        int subsetXORSum(vector<int>& nums) {
            int ans = 0;
            for(int mask = 0; mask < (1 << nums.size()); ++mask){
                int tmp = 0;
                for(int i = 0; i < nums.size(); ++i){
                    if(mask & (1 << i)) tmp ^= nums[i];
                }
                ans += tmp;
            }
            return ans;
        }
    };
    /*
    mask 會有
    0000, 0001 ... 1111
    mask 總共會和 nums.size() 相同大小，例如 4
    而排列組合 0001 是 代表 xor nums[0]
    所以會遇到 0101 就是 xor nums[2] 和 nums[0]
    最後就會把每個排列組合都 xor 到
    */

File: Bit Manipulation/1863. Sum of All Subset XOR Totals.py
# 2024/05/20
class Solution(object):
    def subsetXORSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # bit manipulation (by Editorial)
        # 可以觀察到，按照每個 subset 的加總，也就是把所有數字的 bit 做 or 後的結果，然後在右側補上 N-1 個 0
        # 1:  0 0 1
        # 5:  1 0 1
        # 6:  1 1 0
        # or: 1 1 1
        # ans 28: 1 1 1 0 0
        ans = 0
        N = len(nums)
        for i in range(N):
            ans = ans | nums[i]
        
        # decimal to binary
        bin(ans).replace("0b", "")

        # 右邊補 N-1 個 0，也就是往左移 N-1 個 bit
        ans = ans << N-1

        return ans
        